bin/alpen-cli/src/recovery.rs

```diff
@@ -248,3 +253,83 @@ pub struct InvalidPrivateKey(OneOf<(FromUtf8Error, DescriptorKeyParseError)>);
 #[derive(Debug)]
 #[allow(unused)]
 pub struct InvalidPublicKey(OneOf<(FromUtf8Error, DescriptorKeyParseError)>);
+
+#[derive(Debug)]
+pub struct DescriptorRecoveryKey {
+    pub recover_at: u32,
+    pub desc_string_hash: [u8; 32],
+}
+
+impl DescriptorRecoveryKey {
+    pub fn new(recover_at: u32, desc: &Descriptor<DescriptorPublicKey>) -> Self {
+        let mut hasher = <Sha256 as Digest>::new(); // this is to appease the analyzer
+        hasher.update(desc.to_string().as_bytes());
+        Self {
+            recover_at,
+            desc_string_hash: hasher.finalize().into(),
+        }
+    }
+
+    pub fn encode(&self) -> [u8; 36] {
+        make_buf! {
+            (&self.recover_at.to_be_bytes(), 4),
+            (&self.desc_string_hash, 32)
+        }
+    }
+
+    pub fn decode(bytes: &[u8]) -> Option<Self> {
+        if bytes.len() != 36 {
+            return None;
+        }
+
+        let recover_at = u32::from_be_bytes(unsafe { *(bytes[..4].as_ptr() as *const [_; 4]) });
+        let mut desc_string_hash = [0u8; 32];
+        desc_string_hash.copy_from_slice(&bytes[4..36]);
+
+        Some(Self {
+            recover_at,
+            desc_string_hash,
+        })
+    }
+}
+
+struct BigEndianRangeBounds {
+    start: Bound<[u8; 4]>,
+    end: Bound<[u8; 4]>,
+}
+
+impl BigEndianRangeBounds {
+    fn from_u32_range<R: RangeBounds<u32>>(range: R) -> Self {
+        let start = match range.start_bound() {
+            Bound::Included(&n) => Bound::Included(n.to_be_bytes()),
+            Bound::Excluded(&n) => Bound::Excluded(n.to_be_bytes()),
+            Bound::Unbounded => Bound::Unbounded,
+        };
+
+        let end = match range.end_bound() {
+            Bound::Included(&n) => Bound::Included(n.to_be_bytes()),
+            Bound::Excluded(&n) => Bound::Excluded(n.to_be_bytes()),
+            Bound::Unbounded => Bound::Unbounded,
+        };
+
+        Self { start, end }
+    }
+}
+
+impl RangeBounds<[u8; 4]> for BigEndianRangeBounds {
+    fn start_bound(&self) -> Bound<&[u8; 4]> {
+        match &self.start {
+            Bound::Included(arr) => Bound::Included(arr),
+            Bound::Excluded(arr) => Bound::Excluded(arr),
+            Bound::Unbounded => Bound::Unbounded,
+        }
+    }
+
+    fn end_bound(&self) -> Bound<&[u8; 4]> {
+        match &self.end {
+            Bound::Included(arr) => Bound::Included(arr),
+            Bound::Excluded(arr) => Bound::Excluded(arr),
+            Bound::Unbounded => Bound::Unbounded,
+        }
```

Comment:
As we discussed it would be nice to know why we need this.
Can you please document where appropriate with docstrings or comments?